File: dsa-examples/python/insertion-sort.py
# Insertion Sort
# Time Complexity: Best: O(n), Average: O(n^2), Worst: O(n^2)
def insertion_sort(nums):
    nums = list(nums)
    for i in range(1, len(nums)):
        num = nums[i]
        j = i - 1
        while j > -1 and nums[j] > num:
            nums[j + 1] = nums[j]
            j = j - 1
        nums[j + 1] = num
    return nums

File: dsa-examples/go/insertion-sort.go
package main

import "fmt"

func main() {
	fmt.Println(insertion_sort([]int{5, 2, 9, 1, 4, 3, 8, 6, 7}))
}

func insertion_sort(nums_arr []int) []int {
	nums := nums_arr
	for i := 1; i < len(nums); i++ {
		num := nums[i]
		j := i - 1
		for j > -1 && nums[j] > num {
            nums[j + 1] = nums[j]
            j = j - 1
		}
        nums[j + 1] = num
	}
	return nums
}
